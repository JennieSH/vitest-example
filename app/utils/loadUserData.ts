interface User {
  name: string;
  email: string;
  foods: string[];
  verified: boolean;
  isVip?: boolean;
}

// simulate fetching user data from server
const getUser = async (name: string): Promise<User | undefined> =>
  new Promise((resolve) => {
    const users: User[] = [
      {
        name: "Jennie",
        email: "hello@com.tw",
        foods: ["apple", "banana"],
        verified: true
      },
      {
        name: "Jackson",
        email: "world@com.tw",
        foods: ["kiwi"],
        verified: false
      }
    ];

    setTimeout(() => {
      resolve(users.find((user) => user.name === name));
    }, 1000);
  });

export const updateUserData = async (
  name: string
): Promise<User | undefined> => {
  const user = await getUser(name);

  if (!user) {
    throw new Error("User not found");
  }

  // mutate user object
  user.isVip = name === "Jennie" ? true : false;

  return user;
};
