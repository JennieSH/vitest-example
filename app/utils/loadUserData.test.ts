import { describe, test, expect } from "vitest";
import { updateUserData } from "./loadUserData";

describe("updateUserData", () => {
  test("loads user data as expected", async () => {
    const user = await updateUserData("Jennie");

    // fails the test (show the expected and received values)
    // expect(user).toEqual({});

    expect(user).toEqual({
      name: "Jennie",
      email: "hello@com.tw",
      foods: ["apple", "banana"],
      verified: true,
      isVip: true
    });
  });

  test("throws error when user not found", async () => {
    // since the error is thrown before the expect statement, it cannot be caught.
    // so instead putting the data in the expect statement, should be put the function you're expecting to throw an error
    // const unknownUser = await updateUserData("fakeUser");

    expect(async () => await updateUserData("fakeUser")).rejects.toThrowError(
      "User not found"
    );
  });
});

// describe("updateUserData(Snapshot)", () => {
//   test("loads user data as expected", async () => {
//     const user = await updateUserData("Jennie");

//     // automatically generates a snapshot of the user object
//     // the snapshot is saved in the __snapshots__ folder, and need to be committed to the repository
//      expect(user).toMatchSnapshot();

//     // expect(user).toMatchInlineSnapshot();
//   });
// });

// vi.mock("./updateUserData", () => ({
//   loadUserData: () => ({
//     name: "FakeUser2",
//     email: "hey@com.tw",
//     foods: [],
//     verified: true
//   })
// }));

// describe("updateUserData(mock)", () => {
//   test("loads user data as expected", async () => {
//     const user = await updateUserData("Jennie");

//     expect(user).toEqual({
//       name: "Jennie",
//       email: "hello@com.tw",
//       foods: ["apple", "banana"],
//       verified: true,
//       isVip: true
//     });
//   });
// });
