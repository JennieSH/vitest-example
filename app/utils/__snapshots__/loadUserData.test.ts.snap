// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`updateUserData(Snapshot) > loads user data as expected 1`] = `
{
  "email": "hello@com.tw",
  "foods": [
    "apple",
    "banana",
  ],
  "isVip": true,
  "name": "Jennie",
  "verified": true,
}
`;
